docs/udf-guide.md

```diff
@@ -0,0 +1,159 @@
+# Guide to User-Defined Functions (UDF)
+
+This is a guide to show how to use UDFs in .NET for Apache Spark.
+
+## What are UDFs
+
+[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the vocabulary of Spark SQL’s Domain Specific Language. They transform values from a single row within a table to produce a single corresponding output value per row based on the logic defined in the UDF.
+
+Let's take the following as an example for a UDF definition:
+
+```csharp
+string s1 = "hello";
+Func<Column, Column> udf = Udf<string, string>(
+    str => $"{s1} {str}");
+
+```
+
+And for a sample Dataframe, let's take the following Dataframe `df`:
+
+```text
+    +-------+
+    |   name|
+    +-------+
+    |Michael|
+    |   Andy|
+    | Justin|
+    +-------+
+```
+
+Now let's apply the above defined `udf` to the dataframe `df`:
+
+```csharp
+DataFrame udfResult = df.Select(udf(df["name"]));
+```
+
+This would return the below as the Dataframe `udfResult`:
+
+```text
+    +-------------+
+    |         name|
+    +-------------+
+    |hello Michael|
+    |   hello Andy|
+    | hello Justin|
+    +-------------+
+```
+
+## Good to know while implementing UDFs
+
+One behavior to be aware of while implementing UDFs in .NET for Apache Spark is the way target of the UDF gets serialized. .NET for Apache Spark uses .NET Core which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means:
+
+Sample user code:
```

Comment:
Rather than "Sample user code," perhaps we could have a sentence or two explaining what the following user code does.

Reply:
Done, thanks!